Userland/Test.cpp

```diff
@@ -0,0 +1,517 @@
+        case SymbolicLink:
+            return S_ISLNK(statbuf.st_mode);
+        default:
+            ASSERT_NOT_REACHED();
+        }
+    }
+
+    String m_path;
+    Kind m_kind { Regular };
+};
+
+class HasPermission : public Condition {
+public:
+    enum Permission {
+        Any,
+        Read,
+        Write,
+        Execute,
+    };
+    HasPermission(StringView path, Permission kind)
+        : m_path(path)
+        , m_kind(kind)
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        switch (m_kind) {
+        case Read:
+            return access(m_path.characters(), R_OK) == 0;
+        case Write:
+            return access(m_path.characters(), W_OK) == 0;
+        case Execute:
+            return access(m_path.characters(), X_OK) == 0;
+        case Any:
+            return access(m_path.characters(), F_OK) == 0;
+        default:
+            ASSERT_NOT_REACHED();
+        }
+    }
+
+    String m_path;
+    Permission m_kind { Read };
+};
+
+class StringCompare : public Condition {
+public:
+    enum Mode {
+        Equal,
+        NotEqual,
+    };
+
+    StringCompare(StringView lhs, StringView rhs, Mode mode)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+        , m_mode(mode)
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        if (m_mode == Equal)
+            return m_lhs == m_rhs;
+        return m_lhs != m_rhs;
+    }
+
+    StringView m_lhs;
+    StringView m_rhs;
+    Mode m_mode { Equal };
+};
+
+class NumericCompare : public Condition {
+public:
+    enum Mode {
+        Equal,
+        Greater,
+        GreaterOrEqual,
+        Less,
+        LessOrEqual,
+        NotEqual,
+    };
+
+    NumericCompare(StringView lhs, StringView rhs, Mode mode)
+        : m_mode(mode)
+    {
+        m_lhs = lhs.to_int().value_or(0);
+        m_rhs = rhs.to_int().value_or(0);
+    }
+
+private:
+    virtual bool check() const override
+    {
+        switch (m_mode) {
+        case Equal:
+            return m_lhs == m_rhs;
+        case Greater:
+            return m_lhs > m_rhs;
+        case GreaterOrEqual:
+            return m_lhs >= m_rhs;
+        case Less:
+            return m_lhs < m_rhs;
+        case LessOrEqual:
+            return m_lhs <= m_rhs;
+        case NotEqual:
+            return m_lhs != m_rhs;
+        default:
+            ASSERT_NOT_REACHED();
+        }
+    }
+
+    int m_lhs { 0 };
+    int m_rhs { 0 };
+    Mode m_mode { Equal };
+};
+
+class FileCompare : public Condition {
+public:
+    enum Mode {
+        Same,
+        ModificationTimestampGreater,
+        ModificationTimestampLess,
+    };
+
+    FileCompare(String lhs, String rhs, Mode mode)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+        , m_mode(mode)
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        struct stat statbuf_l;
+        int rc = lstat(m_lhs.characters(), &statbuf_l);
```

Comment:
Are these checks supposed to dereference symlinks? Either way, it needs to be documented in the man page.

Reply:
it _is_ documented "all FILE checks dereference symbolic links".

Reply:
Yes, this is why I'm surprised to see this check *not* dereferencing symlinks, so I thought perhaps that only applies to that section on individual file checks, not on binary checks.

Then we need to fix it here in the code and make it more clear in the man page that this applies to both individual file checks and binary checks.

Reply:
...I think I was too sleepy when I wrote this...
I somehow managed to mix up `lstat` and `stat`...

Reply:
No worries, that's what code review is for 😀